ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+    build.shipwright.io/upload.wait.timeout: 120
+spec: {}
+```
+
+Therefore when `build.shipwright.io/wait.upload` is present the waiting will take place, and by
+expressing `build.shipwright.io/upload.wait.timeout` the actual timeout is configured.
+
+#### Waiting Mechanism
+
+The container-image built by Shipwright is created by a series of steps running sequentially, those
+steps are executed in the build-pod, running each container and inspecting for return-code and
+errors.
+
+The task at hand, waiting for the users' data transfer, will follow the same basic structure, which
+means a simple application meant to wait for a specific condition, and exit gracefully afterwards.
+Here we may keep checking if a given "lock" file exists, when removed the application exists. It also
+must exit with an error on timeout.
+
+Additionally, we can employ this application to make verifications on the streamed data, like for
+instance checksum, ownership (UID/GID) and more.
+
+The implementation written for research and development can be found at [otaviof/waiter][Waiter]
+repository, and it is started by running "`waiter start`". When running it will keep watching the
+file `/var/tmp/waiter.pid`, and when removed, it will exit gracefully. The following example shows it
+running in the container-manager. In the following example the entrypoint receives the argument
+"`start`", and the waiting begins.
+
+```sh
+$ docker run --name=waiter --interactive otaviof/waiter:latest start
+2021/06/01 08:25:24 Saving PID '1' at '/var/tmp/waiter.pid'
+2021/06/01 08:25:24 Attempts '600'...
+2021/06/01 08:25:25 Attempts '599'...
+2021/06/01 08:25:26 Attempts '598'...
+2021/06/01 08:25:27 Done!
+```
+
+When the data transfer is complete, the CLI will execute "`waiter done`", via the same "`kubectl
+exec`" mechanism.
+
+```sh
+$ docker exec waiter waiter done
+2021/06/01 08:25:56 Removing pid-file '/var/tmp/waiter.pid' (1)
+```
+
+This procedure works on the current version of [Tekton (v0.22.0)][TektonV0220] used in Shipwright
+Build-Controller.
+
+##### Waiter on Controller Repository
+
+The `waiter` application will live on the [Build Controller repository][ControllerCmd], side by side
+with other applications, and producing a stand alone container-image. The image must be based on
+UBI8, following the same standards than others.
+
+This image also has `tar` installed, which will be used for the data upload mechanism, as explained
+before.
+
+Another requirement is following the same configuration standard than other external images used by
+the Build Controller, like a environment variable (i.e.: `GIT_CONTAINER_TEMPLATE`) to overwrite the
+image before the controller starts.
+
+At last, it should provide meaningful information for Tekton's `TaskResult`, so we can differentiate
+this step run from others. For the `waiter` process results, it can take advantage of the
+[termination-log pattern][KubernetesTerminationLog] to inform Tekton more detailed execution results.
```

Comment:
Which result(s) do you suggest? Only something like size comes to my mind.

Reply:
On top of my mind, we will inform the possible errors, like permissions, no-space left on device, etc. And possibly things like data-checksum, permissions that changed, and such things.

Reply:
Please look at it as an open channel, so to say, where the `waiter` can inform detailed information about the changes done to the build context.